modules/ding_user/ding_user.module

```diff
@@ -451,58 +451,66 @@ function ding_user_default_authname($name) {
-  _ding_user_update_old_accounts($hash);
-  return _password_base64_encode($hash, drupal_strlen($hash)) . '@ding_user';
+  // Hence the strlen().
+  $authname = _password_base64_encode($hash, strlen($hash)) . '@ding_user';
+
+  _ding_user_update_authname($authname);
+
+  return $authname;
 }
 
 /**
- * This method updates the user authname hash.
+ * Updates authname for existing users.
  *
- * It will become deprecated after a while.
+ * Brings compatibility of user accounts across various php versions.
+ * Issue caused by drupal_strlen() -> mb_strlen(), whose
+ * behavior was changed somewhat in php 5.4+
  *
- * @param string $hash
- *   Hash used as provider users authname.
+ * @see http://php.net/releases/NEWS_5_4_0_beta2.txt
+ * - "Ill-formed UTF-8 check for security enhancements."
  */
-function _ding_user_update_old_accounts($hash) {
-  // Create authname hash using old workflow (the difference is in strlen VS
-  // drupal_strlen).
-  $auth_name_old = _password_base64_encode($hash, strlen($hash)) . '@ding_user';
+function _ding_user_update_authname($authname) {
+  // We'll check for first 32 chars to get a match.
+  // Hope _password_base64_encode() gives unique results.
+  $authname_chunk = substr($authname, 0, 32);
 
-  // Try to find the users uid based on the old hash value (length of the old
-  // hash should be 96).
-  $uid = db_select('authmap', 'a')
+  // Pick the user id that has almost the authname value that we seek,
+  // which is still different from what it should be.
+  $query = db_select('authmap', 'a')
     ->fields('a', array('uid'))
-    ->condition('authname', $auth_name_old)
+    ->condition('authname', $authname_chunk . '%', 'LIKE');
+  $uids = $query
     ->execute()
-    ->fetchField();
-
-  // Proceed only in the case it exists and needs to be updated.
-  if ($uid) {
-    // Create new hash.
-    $auth_name_new = _password_base64_encode($hash, drupal_strlen($hash)) . '@ding_user';
-    $account_new = user_external_load($auth_name_new);
-
-    // Update old account (authmap and init) only when there is no new account
-    // for the same user.
-    if (!$account_new) {
-      db_update('authmap')
-        ->fields(array(
-          'authname' => $auth_name_new,
-        ))
-        ->condition('authname', $auth_name_old)
-        ->execute();
-
-      db_update('users')
-        ->fields(array(
-          'init' => $auth_name_new,
-        ))
-        ->condition('uid', $uid)
-        ->execute();
+    ->fetchCol();
+
+  // In the case there is still a user with requested authname,
+  // update his authname tokens in database.
+  if ($uids) {
+    $latest_uid = max($uids);
+    $original_uid = min($uids);
+
+    // In case of same provider user created erroneously, delete the one
+    // with latest uid assigned. But being aware that there are actually
+    // more users with same authname chunk, yet with different uid's.
+    if ($latest_uid != $original_uid) {
+      user_delete($latest_uid);
```

Comment:
This seems to assume that there will be 2 UIDs matching the chunck at most since we only delete the latest. How can be we sure this is the case?

Also: How can we be sure that the latest uid is the right one to delete?